src/Strategy/Go/Gomod.hs

```diff
@@ -164,14 +198,92 @@ gomodParser = do
+    version :: Parser PackageVersion
+    version = parseSemOrPseudo <|> parseNonCanonical
+      where
+        -- Helpers.
+        semVerText = T.pack <$> lexeme (some (alphaNumChar <|> oneOf ['.', '-', '+', '_', '/']))
+
+        mapLeft _ (Right r) = r
+        mapLeft f (Left l) = f l
+
+        -- Non-canonical versions can be made of any valid string of characters.
+        parseNonCanonical = NonCanonical <$> semVerText
+
+        -- Pseudo-versions are also valid semantic versions. We first determine
+        -- whether a version is properly semantic, and then determine whether it
+        -- could also be a pseudo-version.
+        parseSemOrPseudo = do
+          -- Go modules adds a leading "v". See
+          -- https://golang.org/ref/mod#versions.
+          --
+          -- > Each version starts with the letter v, followed by a semantic
+          -- > version.
+          _ <- char 'v'
+          raw <- semVerText
+
+          -- Once we have the version text, we then try to parse the version
+          -- text as a semver using SemVer.fromText.
+          return $ case SemVer.fromText raw of
+            Right semver -> case reverse $ _versionRelease semver of
+              -- See https://golang.org/ref/mod#pseudo-versions for the forms
+              -- of pseudo-versions.
+              --
+              -- Pseudo-versions always have a trailing "pseudo" pre-release
+              -- identifier. If there are other pre-release identifiers, then
+              -- there is a leading pre-release identifier of 0.
+              --
+              -- When a semantic version matches a possible pseudo-version form,
+              -- we then check whether the pre-release identifiers are formatted
+              -- as if the version is a pseudo-version. If the parse fails, then
+              -- we treat this version as a normal semantic version.
+              [IText preReleaseId] -> mapLeft (const $ Semantic semver) $ parsePseudoPreRelease preReleaseId
+              (IText preReleaseId) : (INum 0) : _ -> mapLeft (const $ Semantic semver) $ parsePseudoPreRelease preReleaseId
+
+              -- If the semantic version has no pre-release identifiers, it
+              -- cannot possibly be a pseudo-version.
+              _ -> Semantic semver
+            Left _ -> NonCanonical raw
+
+        -- Here, we run a sub-parser to determine whether the "pseudo"
+        -- pre-release identifier is correctly formatted. This is always of the
+        -- form "yyyymmddhhmmss-abcdefabcdef" - a timestamp, and then a commit
+        -- hash.
+        --
+        -- For the backend's sake, we want to upload the commit hash (which
+        -- identifies the version of the dependency).
+        --
+        -- See https://golang.org/ref/mod#pseudo-versions for more details.
+        parsePseudoPreRelease = parse parser ""
+          where
+            parser :: Parser PackageVersion
+            parser = do
```

Comment:
Done, but what's the reasoning for this convention? This actually feels _less_ readable to me.

Reply:
TLDR: Applicative syntax requires less reasoning, and more reading.

`do` notation allows you to declare variables which are the result of monadic computations, and then use them anywhere.  While that is quite useful and sometimes absolutely necessary, it takes more effort to reason about what can happen in the code, since we now have the full power of monads at our disposal.

In applicative syntax, computational power is limited, and therefore can have fewer possible effects.

We also have not been using do-notation for parsers in most of the project, and the style was intentionally chosen, despite the drawback of readability.  We should continue the pattern until we can agree that it should be changed.

It is true that applicative syntax is unarguably harder to read, though I would say not by much once you understand the functor/applicative combinator symbols `<$ $> <$> <* *> <*>`.

Reply:
Yeah, I buy the "use less power" argument.